Add Bazel Docs (#160)
bufbuild/docs.buf.build: docs/build-systems/bazel.md

* [Lint] Protobuf sources using the [`buf_lint_test`](#buf-lint-test) rule.
* Perform [breaking change detection][breaking] for Protobuf [Inputs] using the [`buf_breaking_test`](#buf-breaking-test) rule.
* Use the [`buf`][buf_cli] CLI as a Bazel [toolchain](#toolchains).
* Use the [Gazelle](#gazelle) extension to generate Bazel rules.

## Setup

To get started, you need to add a series of imports to your Bazel `WORKSPACE`, replacing the `<SHA256>` and `<VERSION>` with values from a [specific `rules_buf` release][release]:

```python title="WORKSPACE"
# Prerequisites
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "rules_buf",
    sha256 = "<SHA256>",
    urls = [        
        "https://github.com/bufbuild/rules_buf/releases/download/<VERSION>/rules_buf-<VERSION>.zip",
    ],
)

# Buf rules
load("@rules_buf//buf:repositories.bzl", "rules_buf_dependencies", "rules_buf_toolchains")

rules_buf_dependencies()

rules_buf_toolchains()

# Protobuf rules
load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")

rules_proto_dependencies()

rules_proto_toolchains()
```

> ### Using a specific version of the Protobuf rules

Reviewer: You likely want to change this from "Protobuf rules" to "rules_proto", I read this as setting a specific version of "rules_buf", which I thought was covered in the previous section.

# Prerequisites
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

# Load rules_proto version 4.0.0
http_archive(
    name = "rules_proto",
    sha256 = "66bfdf8782796239d3875d37e7de19b1d94301e8972b3cbd2446b332429b4df1",
    strip_prefix = "rules_proto-4.0.0",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/rules_proto/archive/refs/tags/4.0.0.tar.gz",
        "https://github.com/bazelbuild/rules_proto/archive/refs/tags/4.0.0.tar.gz",
    ],
)

load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")

rules_proto_dependencies()

rules_proto_toolchains()

# Load rules_buf version 0.1.0
http_archive(
    name = "rules_buf",
    sha256 = "d6b2513456fe2229811da7eb67a444be7785f5323c6708b38d851d2b51e54d83",
    urls = [        
         "https://github.com/bufbuild/rules_buf/releases/download/v0.1.0/rules_buf-v0.1.0.zip",
    ],
)

# Load the other rules_buf assets
```

### Pinning the `buf` cli version

By default, `rules_buf_toolchains` loads the latest buf version. For hermetic builds pin the cli version using the version attribute.

```python title="WORKSPACE" {1-2}
-rules_buf_toolchains()

> +rules_buf_toolchains(version = "v1.5.0")

Author: Any version of `rules_buf` can download any version of `buf`

```

## Rules

The rules work alongside `proto_library` rules. You can configure `rules_buf` using a [`buf.yaml`][buf_yaml] configuration file. Export the `buf.yaml` using `exports_files(["buf.yaml"])` to reference it. For repositories that contain a [`buf.work.yaml`][buf_work_yaml] that references multiple `buf.yaml` files, you need to export and reference each `buf.yaml` file independently.

> We recommend using the [Gazelle extension](#gazelle) to generate the following rules.

### `buf_lint_test` {#buf-lint-test}

`buf_lint_test` is a test rule that lints one or more `proto_library` targets.

#### Example

```python
load("@rules_buf//buf:defs.bzl", "buf_lint_test")
load("@rules_proto//proto:defs.bzl", "proto_library")

proto_library(
    name = "foo_proto",
    srcs = ["pet.proto"],
    deps = ["@go_googleapis//google/type:datetime_proto"],
)

buf_lint_test(
    name = "foo_proto_lint",    
    targets = [":foo_proto"],
    config = "buf.yaml",
)
```

This can be run as:

```sh
$ bazel test :foo_proto_lint
```

We recommend having a single `buf_lint_test` for each `proto_library` target. The [Gazelle extension](#gazelle) can generate them in the same pattern.

> #### Attributes

Reviewer: I would probably put attributes before example

[//]: # (The table is copied from documentation generated by stardoc: https://github.com/bufbuild/rules_buf/blob/main/buf/README.md )

Name  | Description | Type | Mandatory | Default
:-----| :-----------| :--- | :-------- | :------
`name` |  A unique name for this target.   | [Name][config_name] | required |  |
`config` |  The [`buf.yaml`][buf_yaml] file. | [Label][config_label] | optional | None |
`targets` |  `proto_library` targets to lint | [List of labels][config_label] | required |  |

> ### `buf_breaking_test` {#buf-breaking-test}

Reviewer: `buf_lint_test` has `#### Example` but this section does not, we should be consistent

> `buf_breaking_test` is a test rule that checks one or more `proto_library` targets for breaking changes. It requires an [image](/reference/images) file to check against. Here's an example usage of the rule:

Reviewer: We should discuss the requirement to have an against image.

```python
load("@rules_buf//buf:defs.bzl", "buf_breaking_test")
load("@rules_proto//proto:defs.bzl", "proto_library")

proto_library(
    name = "foo_proto",
    srcs = ["foo.proto"],
)

buf_breaking_test(
    name = "foo_proto_breaking",
    against = "//:image.bin", # The Image file to check against.
    targets = [":foo_proto"], # The Protobuf library
    config = ":buf.yaml",
)
```

This can be run as:

```terminal
$ bazel test :foo_proto_breaking
```

We recommend having a single `buf_breaking_test` for each Buf module. For repositories that contain a `buf.work.yaml` that references multiple `buf.yaml` files, there needs to be exactly one `buf_breaking_test` for each `buf.yaml` file.

> Alternatively, a single `buf_breaking_test` can be used against each `proto_library` target. For this to work, `limit_to_input_files` attribute must be set to `True` as the against image file may contain other Protobuf files. Although this is closer to how bazel operates, for this particular use case it is not recommended. This [section](#example-module-vs-package-mode) explains the differences with an example.

Reviewer 2: ```suggestion
Alternatively, a single `buf_breaking_test` can be used against each `proto_library` target. For this to work, `limit_to_input_files` attribute must be set to `True` as the against image file may contain other Protobuf files. Although this is closer to how bazel operates, for this particular use case it is not recommended. See the [module vs package mode example](#example-module-vs-package-mode) for a concrete example of the differences.
```

I'm generally not a fan of "see [this]" thing where the this doesnt really indicate where it goes.

rules_proto_dependencies()

rules_proto_toolchains()
```

### Using a specific version of the `rules_proto`

[`rules_proto`][rules_proto] is required to use `rules_buf`. By default, `rules_buf` automatically loads `rules_proto`, but you can use a specific version of it by loading it _before_ `rules_buf`. Here's an example:

```python title="WORKSPACE"
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "rules_proto",
    sha256 = "66bfdf8782796239d3875d37e7de19b1d94301e8972b3cbd2446b332429b4df1",
    strip_prefix = "rules_proto-4.0.0",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/rules_proto/archive/refs/tags/4.0.0.tar.gz",
        "https://github.com/bazelbuild/rules_proto/archive/refs/tags/4.0.0.tar.gz",
    ],
)

load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")

rules_proto_dependencies()

rules_proto_toolchains()

http_archive(
    name = "rules_buf",
    sha256 = "d6b2513456fe2229811da7eb67a444be7785f5323c6708b38d851d2b51e54d83",
    urls = [        
         "https://github.com/bufbuild/rules_buf/releases/download/v0.1.0/rules_buf-v0.1.0.zip",

>     ],

Reviewer 2: Here we supply a sha256 and url, but line 21-25 have placeholders deferring to the readme. We should either have placeholders in both places, or keep the latest version around here and have the release instructions for rules_buf update these docs.

Author: Changing this to placeholders

)

# Load the other rules_buf assets
```

### Pinning the `buf` version

By default, `rules_buf_toolchains` loads the latest `buf` version. For hermetic builds pin the `buf` version using the version attribute.

```python title="WORKSPACE" {1-2}
# rules_buf fetches the sha based on the version number, the version is enough for hermetic builds.
-rules_buf_toolchains()
+rules_buf_toolchains(version = "v1.5.0")
```

## Rules

The rules work alongside `proto_library` rules. You can configure `rules_buf` using a [`buf.yaml`][buf_yaml] configuration file. Export the `buf.yaml` using `exports_files(["buf.yaml"])` to reference it. For repositories that contain a [`buf.work.yaml`][buf_work_yaml] that references multiple `buf.yaml` files, you need to export and reference each `buf.yaml` file independently.

> We recommend using the [Gazelle extension](#gazelle) to generate the following rules.

### `buf_dependencies` {#buf-dependencies}

> `buf_dependencies` is a [repository rule][repository_rule] that downloads one or more modules from the [BSR][bsr] and generates build files using Gazelle. [Setup Gazelle][gazelle_setup] to use this rule. To also use Gazelle to generate this rule see the [Dependencies](#gazelle-dependencies) section.

Reviewer 2: Don't you need our gazelle extension to patch up `deps` automatically in `proto_library` rules?

> `buf_dependencies` is a [repository rule][repository_rule] that downloads one or more modules from the [BSR][bsr] and generates build files using Gazelle. [Setup Gazelle][gazelle_setup] to use this rule. To also use Gazelle to generate this rule see the [Dependencies](#gazelle-dependencies) section.

Author: Only if they use gazelle for generating `proto_library` rules

#### Example

```python title="WORKSPACE"
load("@rules_buf//buf:defs.bzl", "buf_dependencies")

buf_dependencies(
    name = "buf_deps",
    modules = [
        "buf.build/envoyproxy/protoc-gen-validate:dc09a417d27241f7b069feae2cd74a0e", 
        "buf.build/acme/petapis:84a33a06f0954823a6f2a089fb1bb82e",
    ],
)
```

```python title="BUILD"
load("@rules_proto//proto:defs.bzl", "proto_library")

# imports "validate/validate.proto"
proto_library(
    name = "foo_proto",
    srcs = ["pet.proto"],    
    deps = ["@buf_deps//validate:validate_proto"],
)
```

We recomment using a single `buf_dependencies` rule for each `buf.yaml` file. The [Gazelle extension](#gazelle-dependencies) does this by default.

### `buf_lint_test` {#buf-lint-test}

`buf_lint_test` is a test rule that lints one or more `proto_library` targets.

#### Attributes

[//]: # (The table is copied from documentation generated by stardoc: https://github.com/bufbuild/rules_buf/blob/main/buf/README.md )

Name  | Description | Type | Mandatory | Default
:-----| :-----------| :--- | :-------- | :------
`name` |  A unique name for this target.   | [Name][config_name] | required |  |

> `config` |  The [`buf.yaml`][buf_yaml] file. | [Label][config_label] | optional | None |

Reviewer 2: Lets note that if no config is specified, the default buf config will be applied (is that the same as you get from `buf mod init` or is there some other default?)

Author: That's the default one I'll add a note here

```

Now run Gazelle `update-repos` command
```terminal
$ bazel run //:gazelle-update-repos
```

This will create the file `buf_deps.bzl` with the `buf_deps` macro that loads the `buf_dependencies` rules. It will also call the macro from the `WORKSPACE` file.

#### Arguments

Argument  | Mandatory | Default
:-----|:--- | :--- 
`-from_file file` <br/><br/> Must be one of `buf.work.yaml`, `buf.yaml`, `buf.lock`. When using `buf.work.yaml` or `buf.yaml` the rule will import from the associated `buf.lock` file(s). | required |  
`-to_macro macroFile%defName` <br/><br/> Tells Gazelle to write new repository rules into a `.bzl` macro function rather than the `WORKSPACE` file. | optional | ''
<code>-prune true&#124;false</code> <br/><br/> When true, Gazelle will remove `buf_dependencies` rules that no longer have equivalent `buf.yaml` files. | optional | `false`

### Lint

By default, a `buf_lint_test` rule is generated for each of the `proto_library` rule generated by Gazelle. It will pick up the `buf.yaml` that the Protobuf package belongs to.

Run this command to list the generated lint rules:

```terminal
$ bazel query 'kind(buf_lint_test, //...)'
```

### Breaking change detection

> To run [breaking change detection][breaking] against Protobuf sources, you need to add a Gazelle [directive][gazelle_directive] that points to an [Image][image] target to generate breaking change detection rules. Gazelle directives are top-level commits in Bazel [`BUILD` files][build_files] that provide Gazelle with configuration.

Reviewer 2: ```suggestion
To run [breaking change detection][breaking] against Protobuf sources, you need to add a Gazelle [directive][gazelle_directive] that points to an [Image][image] target to generate breaking change detection rules. Gazelle directives are top-level comments in Bazel [`BUILD` files][build_files] that provide Gazelle with configuration.
```

#### Attributes

| Name  | Description | Type | Mandatory | Default |
| :------------- | :------------- | :------------- | :------------- | :------------- |
| name |  A unique name for this repository.   | [Name][config_name] | required |  |
| modules |  The module pins `remote/owner/repo:revision`   | List of strings | required |  |

#### Example

```python title="WORKSPACE"
load("@rules_buf//buf:defs.bzl", "buf_dependencies")

buf_dependencies(
    name = "buf_deps",
    modules = [
        "buf.build/envoyproxy/protoc-gen-validate:dc09a417d27241f7b069feae2cd74a0e", 
        "buf.build/acme/petapis:84a33a06f0954823a6f2a089fb1bb82e",
    ],
)
```

```python title="BUILD"
load("@rules_proto//proto:defs.bzl", "proto_library")

# imports "validate/validate.proto"
proto_library(
    name = "foo_proto",
    srcs = ["pet.proto"],    
    deps = ["@buf_deps//validate:validate_proto"],
)
```

> We recomment using a single `buf_dependencies` rule for each `buf.yaml` file. The [Gazelle extension](#gazelle-dependencies) does this by default.

Reviewer 3: ```suggestion
We recommend using a single `buf_dependencies` rule for each `buf.yaml` file. The [Gazelle extension](#gazelle-dependencies) does this by default.
```

Start by [setting up](#overview) `rules_buf`, then set up Gazelle using the [official instructions][gazelle_setup].

Once Gazelle is set up, add the following snippet at the end of the `WORKSPACE` file:

```python title="WORKSPACE"
load("@rules_buf//gazelle/buf:repositories.bzl", "gazelle_buf_dependencies")

gazelle_buf_dependencies()
```

Now modify the `BUILD` file with the `gazelle` target to include the `buf` extension:

```python title="BUILD" {1-2,4-13,17}
-load("@bazel_gazelle//:def.bzl", "gazelle")
+load("@bazel_gazelle//:def.bzl", "gazelle", "gazelle_binary")

+gazelle_binary(
+    name = "gazelle-buf",    
+    languages = [
+        # Loads the native proto extension
+        "@bazel_gazelle//language/proto:go_default_library"
+        # Loads the Buf extension
+        "@rules_buf//gazelle/buf:buf",
+        # NOTE: This needs to be loaded after the proto language
+    ],
+)

gazelle(
    name = "gazelle",
+    gazelle = ":gazelle-buf",
)
```

Export the `buf.yaml` file by adding `exports_files(["buf.yaml"])` to the `BUILD` file.

> Inside Buf [workspaces][workspaces], make sure to export each `buf.yaml` file.

Now run Gazelle
```terminal
$ bazel run //:gazelle
```

> This will now take care of updating your proto build files, just run `//:gazelle` whenever proto files are added/removed.

Reviewer 3: ```suggestion
This will take care of updating your proto build files, just run `//:gazelle` whenever proto files are added/removed.
```

* [Lint][lint] Protobuf sources using the [`buf_lint_test`](#buf-lint-test) rule.
* Perform [breaking change detection][breaking] for Protobuf [Inputs][inputs] using the [`buf_breaking_test`](#buf-breaking-test) rule.
* Use the [`buf`][buf_cli] CLI as a Bazel [toolchain](#toolchains).
* Use the [Gazelle](#gazelle) extension to generate Bazel rules.

## Setup

To get started, you need to add a series of imports to your Bazel `WORKSPACE`, replacing the `<SHA256>` and `<VERSION>` with values from a [specific `rules_buf` release][release]:

```python title="WORKSPACE"
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "rules_buf",
    sha256 = "<SHA256>",
    urls = [        
        "https://github.com/bufbuild/rules_buf/releases/download/<VERSION>/rules_buf-<VERSION>.zip",
    ],
)

load("@rules_buf//buf:repositories.bzl", "rules_buf_dependencies", "rules_buf_toolchains")

rules_buf_dependencies()

rules_buf_toolchains()

load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")

rules_proto_dependencies()

rules_proto_toolchains()
```

### Using a specific version of the `rules_proto`

> [`rules_proto`][rules_proto] is required to use `rules_buf`. By default, `rules_buf` automatically loads `rules_proto`, but you can use a specific version of it by loading it _before_ `rules_buf`. Here's an example:

Reviewer 4: Might want to reword this to say:

> By default, `rules_buf` automatically loads the `rules_proto` `http_archive` so you don't need to specify that in your WORKSPACE file.  However, you can use a specific version of `rules_proto` if you like by loading the `http_archive` explicitly _before_ `rules_buf`.  Here's an example:

```

We recommend having a single `buf_lint_test` for each `proto_library` target. The [Gazelle extension](#gazelle) can generate them in the same pattern.

### `buf_breaking_test` {#buf-breaking-test}

`buf_breaking_test` is a test rule that checks one or more `proto_library` targets for breaking changes. It requires an [image](/reference/images) file to check against.

#### Attributes
[//]: # (The table is copied from documentation generated by stardoc: https://github.com/bufbuild/rules_buf/blob/main/buf/README.md )

Name  | Description | Type | Mandatory | Default
:-----|:----------- | :--- | :-------- | :------
`name` |  A unique name for this target.   | [Name][config_name] | required |  |
`against` |  The image file to check against.  | [Label][config_label] | required |  |
`config` |  The `buf.yaml` file.  | [Label][config_label] | optional | `None` |
`exclude_imports` |  Exclude imports from breaking change detection.  | Boolean | optional | `False` |
`limit_to_input_files` |  Only run breaking checks against the files in the targets. This has the effect of filtering the against image to only contain the files in the input. | Boolean | optional | `True` |

> `targets`|  `proto_library` targets to check for breaking changes | [List of labels][config_label] | `[]` |

Reviewer 4: Is `targets` required or optional?  In the docs, the `[]` brackets show under the `Mandatory` header and `Defaults` is blank.

#### Example

```python
load("@rules_buf//buf:defs.bzl", "buf_breaking_test")
load("@rules_proto//proto:defs.bzl", "proto_library")

proto_library(
    name = "foo_proto",
    srcs = ["foo.proto"],
)

buf_breaking_test(
    name = "foo_proto_breaking",
    against = "//:image.bin", # The Image file to check against.
    targets = [":foo_proto"], # The Protobuf library
    config = ":buf.yaml",
)
```

This can be run as:

```terminal
$ bazel test :foo_proto_breaking
```

We recommend having a single `buf_breaking_test` for each `buf.yaml`. For repositories that contain a `buf.work.yaml` that references multiple `buf.yaml` files, there needs to be exactly one `buf_breaking_test` for each `buf.yaml` file.

> Alternatively, a single `buf_breaking_test` can be used against each `proto_library` target. For this to work, `limit_to_input_files` attribute must be set to `True` as the against image file may contain other Protobuf files. Although this is closer to how bazel operates, for this particular use case it is not recommended. See the [module vs package mode example](#example-module-vs-package-mode) for a concrete example of the differences.

Reviewer 4: Maybe format against as `against` (second line 'must be set to `True` as the `against` image)

> Alternatively, a single `buf_breaking_test` can be used against each `proto_library` target. For this to work, `limit_to_input_files` attribute must be set to `True` as the against image file may contain other Protobuf files. Although this is closer to how bazel operates, for this particular use case it is not recommended. See the [module vs package mode example](#example-module-vs-package-mode) for a concrete example of the differences.

Reviewer 4: Capitalize Bazel (second line 'closer to how Bazel operates')

The [Gazelle extension](#gazelle) can generate `buf_breaking_test` in either levels of granularity.

#### Image inputs {#image-input}

You can generate an [Image][image] file like this:

```terminal

> $ buf build --exclude-imports -o image.bin <input>

Reviewer 4: Dumb nit, so feel free to ignore me, but maybe make this:

> The above rules can also use image files as inputs.  You can use the Buf CLI to generate an image file like this:

It was just kind of odd to me to see the CLI command here because all other commands seem to be Bazel-specific and then this was a one-off using the CLI.  It just makes it more clear to me /shrug.

```

The `<input>` is often a directory containing a `buf.yaml`, but all of the other [Input formats](/reference/inputs) are also supported.

We recommend storing the Image file in a `testdata` directory and checking it in to version control and updating it as needed. In the case of repositories that follow a versioning scheme like [semver][semver], you can update it on each new release either manually or with a post-release hook.

As an alternative to checking the Image file into version control, you can use CI artifacts. Many CI servers, like [Travis CI][travis], enable you to upload build artifacts to a backend like [S3][s3]. In CI, you can set up a pipeline to build the Image on each commit and then add those artifacts to your `WORKSPACE`:

```python title="WORKSPACE"
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")

> # Assuming your using s3 and bucket is at http://s3-us-east-1.amazonaws.com/bucket/foo/bar 

Reviewer 4: `you're`

Start by [setting up](#overview) `rules_buf`, then set up Gazelle using the [official instructions][gazelle_setup].

Once Gazelle is set up, add the following snippet at the end of the `WORKSPACE` file:

```python title="WORKSPACE"
load("@rules_buf//gazelle/buf:repositories.bzl", "gazelle_buf_dependencies")

gazelle_buf_dependencies()
```

Now modify the `BUILD` file with the `gazelle` target to include the `buf` extension:

```python title="BUILD" {1-2,4-13,17}
-load("@bazel_gazelle//:def.bzl", "gazelle")
+load("@bazel_gazelle//:def.bzl", "gazelle", "gazelle_binary")

+gazelle_binary(
+    name = "gazelle-buf",    
+    languages = [
+        # Loads the native proto extension
+        "@bazel_gazelle//language/proto:go_default_library"
+        # Loads the Buf extension
+        "@rules_buf//gazelle/buf:buf",
+        # NOTE: This needs to be loaded after the proto language
+    ],
+)

gazelle(
    name = "gazelle",
+    gazelle = ":gazelle-buf",
)
```

Export the `buf.yaml` file by adding `exports_files(["buf.yaml"])` to the `BUILD` file.

> Inside Buf [workspaces][workspaces], make sure to export each `buf.yaml` file.

Now run Gazelle
```terminal
$ bazel run //:gazelle
```

> The buf gazelle extension will now take care of updating your proto build files, just run `//:gazelle` whenever proto files are added/removed.

Reviewer 4: buf gazelle needs some sort of formatting or capitalization, but not sure what.

### Dependencies {#gazelle-dependencies}

Gazelle can also be used to generate `buf_dependencies` rules. It imports dependencies from `buf.lock` files.

Add the following code to the `BUILD` file,

```python title="BUILD"
gazelle(
    name = "gazelle-update-repos",
    args = [
        # This can also be `buf.yaml` and `buf.lock`.
        "--from_file=buf.work.yaml",
        # This is optional but recommended, if absent gazelle
        # will add the rules directly to WORKSPACE
        "-to_macro=buf_deps.bzl%buf_deps",
        # Deletes outdated repo rules
        "-prune",
    ],
    command = "update-repos",
    gazelle = ":gazelle-buf",
)
```

> Now run Gazelle update-repos command

Reviewer 4: Now run the Gazelle `update-repos` command

```terminal
$ bazel run //:gazelle-update-repos
```

This will create the file `buf_deps.bzl` with the `buf_deps` macro that loads the `buf_dependencies` rules. It will also call the macro from the `WORKSPACE` file.

#### Arguments

Argument  | Mandatory | Default
:-----|:--- | :--- 
`-from_file file` <br/><br/> Must be one of `buf.work.yaml`, `buf.yaml`, `buf.lock`. When using `buf.work.yaml` or `buf.yaml` the rule will import from the associated `buf.lock` file(s). | required |  

> `-to_macro macroFile%defName` <br/><br/> Tells Gazelle to write new repository rules into a `.bzl` macro function rather than the `WORKSPACE` file. | optional | 

Reviewer 4: I'm assuming the default value for this is an empty string?  May want to specify that.

> Scroll to the section on [Image inputs](#image-input) for instructions on maintaining Image files themselves.

Add this Gazelle directive to the top of the `BUILD` file at the root of the [Buf Module][module], which is the directory with a [`buf.yaml`][buf_yaml] file.

```python title="BUILD"
# gazelle:buf_breaking_against //:against_image_file
```

You can generate `buf_breaking_test` in two different modes: [module mode](#module-mode) (preferred) and [package mode](#package-mode).

#### Module mode (preferred) {#module-mode}

This is the default and preferred mode. `buf_breaking_test` is generated for each buf module. The rule will reference all the `proto_library` rules that are part of a buf module. This way the test can detect if any files are deleted.

Once the `buf_breaking_against` directive is added, run `gazelle`:

```terminal
$ bazel run //:gazelle
```

Run this command to list the generated breaking rules:

```terminal
$ bazel query 'kind(buf_breaking_test, //...)'
```

> This mimics running `buf breaking` on a module. This is the most accurate way to check for breaking changes. However, depending on multiple targets at once is an anti-pattern in bazel, so that's why we've provided package mode as an alternative.

Reviewer 4: Capitalize Bazel

    sha256 = "66bfdf8782796239d3875d37e7de19b1d94301e8972b3cbd2446b332429b4df1",
    strip_prefix = "rules_proto-4.0.0",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/rules_proto/archive/refs/tags/4.0.0.tar.gz",
        "https://github.com/bazelbuild/rules_proto/archive/refs/tags/4.0.0.tar.gz",
    ],
)

load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")

rules_proto_dependencies()

rules_proto_toolchains()

# Load rules_buf version 0.1.0
http_archive(
    name = "rules_buf",
    sha256 = "d6b2513456fe2229811da7eb67a444be7785f5323c6708b38d851d2b51e54d83",
    urls = [        
         "https://github.com/bufbuild/rules_buf/releases/download/v0.1.0/rules_buf-v0.1.0.zip",
    ],
)

# Load the other rules_buf assets
```

### Pinning the `buf` cli version

By default, `rules_buf_toolchains` loads the latest buf version. For hermetic builds pin the cli version using the version attribute.

```python title="WORKSPACE" {1-2}
# rules_buf fetches the sha based on the version number, the version is enough for hermetic builds.
-rules_buf_toolchains()
+rules_buf_toolchains(version = "v1.5.0")
```

## Rules

> The rules work alongside `proto_library` rules. You can configure `rules_buf` using a [`buf.yaml`][buf_yaml] configuration file. Export the `buf.yaml` using `exports_files(["buf.yaml"])` to reference it. For repositories that contain a [`buf.work.yaml`][buf_work_yaml] that references multiple `buf.yaml` files, you need to export and reference each `buf.yaml` file independently.

Author: I think we should add a link to `export_files` docs and we can also point them to the [examples](https://github.com/bufbuild/rules_buf/tree/main/examples) in the repo

> We recommend using the [Gazelle extension](#gazelle) to generate the following rules.

### `buf_dependencies` {#buf-dependencies}

> `buf_dependencies` is a repository rule that downloads one or more modules from the [BSR][bsr] and generates build files using Gazelle. [Setup Gazelle][gazelle_setup] to use this rule. To also use Gazelle to generate this rule see the [Dependencies](#gazelle-dependencies) section.

Author: The later part is true though, to hand write and use the rule you need gazelle because the rule runs `gazelle` to generate `proto_library` targets.
They can also use `gazelle` to generate this rule.

:-----|:----------- | :--- | :-------- | :------
`name` |  A unique name for this target.   | [Name][config_name] | required |  |
`against` |  The image file to check against.  | [Label][config_label] | required |  |
`config` |  The `buf.yaml` file.  | [Label][config_label] | optional | `None` |
`exclude_imports` |  Exclude imports from breaking change detection.  | Boolean | optional | `False` |
`limit_to_input_files` |  Only run breaking checks against the files in the targets. This has the effect of filtering the against image to only contain the files in the input. | Boolean | optional | `True` |
`targets`|  `proto_library` targets to check for breaking changes | [List of labels][config_label] | `[]` |

#### Example

```python
load("@rules_buf//buf:defs.bzl", "buf_breaking_test")
load("@rules_proto//proto:defs.bzl", "proto_library")

proto_library(
    name = "foo_proto",
    srcs = ["foo.proto"],
)

buf_breaking_test(
    name = "foo_proto_breaking",
    against = "//:image.bin", # The Image file to check against.
    targets = [":foo_proto"], # The Protobuf library
    config = ":buf.yaml",
)
```

This can be run as:

```terminal
$ bazel test :foo_proto_breaking
```

> We recommend having a single `buf_breaking_test` for each Buf module. For repositories that contain a `buf.work.yaml` that references multiple `buf.yaml` files, there needs to be exactly one `buf_breaking_test` for each `buf.yaml` file.

Reviewer 3: Since we reference the `buf.yaml` immediately after, this might be more clear if we drop "Buf module" and replace it with `buf.yaml`:

```suggestion
We recommend having a single `buf_breaking_test` for each `buf.yaml`. For repositories that contain a `buf.work.yaml` that references multiple `buf.yaml` files, there needs to be exactly one `buf_breaking_test` for each `buf.yaml` file.
```

Alternatively, a single `buf_breaking_test` can be used against each `proto_library` target. For this to work, `limit_to_input_files` attribute must be set to `True` as the against image file may contain other Protobuf files. Although this is closer to how bazel operates, for this particular use case it is not recommended. See the [module vs package mode example](#example-module-vs-package-mode) for a concrete example of the differences.

The [Gazelle extension](#gazelle) can generate `buf_breaking_test` in either levels of granularity.

#### Against image

You can generate an [Image] file like this:

```terminal
$ buf build --exclude-imports -o image.bin <input>
```

> The `<input>` can be a path to a Buf module or in one of [several other Image formats](/reference/inputs).

Reviewer 3: Given that we're talking about generic inputs here, we could rephrase this a bit to make it a little more clear. Something along the lines of the following:

```suggestion
The `<input>` is often a directory containing a `buf.yaml`, but all of the other [Input formats](/reference/inputs) are also supported.
```

Add the following code to the `BUILD` file,

```python title="BUILD"
gazelle(
    name = "gazelle-update-repos",
    args = [
        # This can also be `buf.yaml` and `buf.lock`.
        "--from_file=buf.work.yaml",
        # This is optional but recommended, if absent gazelle
        # will add the rules directly to WORKSPACE
        "-to_macro=buf_deps.bzl%buf_deps",
        # Deletes outdated repo rules
        "-prune",
    ],
    command = "update-repos",
    gazelle = ":gazelle-buf",
)
```

Now run Gazelle update-repos command
```terminal
$ bazel run //:gazelle-update-repos
```

This will create the file `buf_deps.bzl` with the `buf_deps` macro that loads the `buf_dependencies` rules. It will also call the macro from the `WORKSPACE` file.

#### Arguments

Argument  | Mandatory | Default
:-----|:--- | :--- 
`-from_file file` <br/><br/> Must be one of `buf.work.yaml`, `buf.yaml`, `buf.lock`. When using `buf.work.yaml` or `buf.yaml` the rule will import from the associated `buf.lock` file(s). | required |  
`-to_macro macroFile%defName` <br/><br/> Tells Gazelle to write new repository rules into a `.bzl` macro function rather than the `WORKSPACE` file. | optional | 

> `-prune true&#124;false` <br/><br/> When true, Gazelle will remove `buf_dependencies` rules that no longer have equivalent `buf.yaml` files. | optional | `false`

Author: `|`

> We recommend using the [Gazelle extension](#gazelle) to generate the following rules.

### `buf_dependencies` {#buf-dependencies}

> `buf_dependencies` is a repository rule that downloads one or more modules from the [BSR][bsr] and generates build files using Gazelle. [Setup Gazelle][gazelle_setup] to use this rule. To also use Gazelle to generate this rule see the [Dependencies](#gazelle-dependencies) section.

Reviewer: What actually happens here? Are these vendored? How do they stay up to date? This should all be documented

Author: > Are these vendored? How do they stay up to date?

As a bazel user doesn't the repository rule part answer these?
